File: log_filter.rb
# log_filter.rb
#!/usr/bin/env ruby
require 'json'
require 'optparse'
require 'date'
require 'time'

CONFIG_FILE = 'log_filter_config.json'

COLORS = {
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  reset: "\e[0m"
}

LEVEL_COLORS = {
  'ERROR' => :red,
  'WARN' => :yellow,
  'INFO' => :green,
  'DEBUG' => :cyan
}

class LogFilter
  attr_reader :config

  def initialize
    @config = load_config
  end

  def load_config
    if File.exist?(CONFIG_FILE)
      JSON.parse(File.read(CONFIG_FILE))
    else
      { 'filters' => [], 'levels' => [], 'since' => '1h' }
    end
  end

  def save_config
    File.write(CONFIG_FILE, JSON.pretty_generate(@config))
  end

  def parse_since(since_str)
    return 3600 unless since_str
    num = since_str.to_i
    unit = since_str[-1]
    case unit
    when 'm' then num * 60
    when 'h' then num * 3600
    when 'd' then num * 86400
    else 3600
    end
  end

  def parse_log_line(line)
    patterns = [
      /\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\]\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)/,
      /(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)/,
      /(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)/,
      /^(ERROR|WARN|INFO|DEBUG)\s+(.*)/
    ]
    patterns.each do |pattern|
      match = line.match(pattern)
      if match
        if match.length == 4
          ts = DateTime.parse(match[1]) rescue nil
          return { timestamp: ts, level: match[2], message: match[3], raw: line }
        elsif match.length == 3
          return { timestamp: nil, level: match[1], message: match[2], raw: line }
        end
      end
    end
    { timestamp: nil, level: nil, message: line, raw: line }
  end

  def highlight(text, filter, regex_pattern)
    if filter
      text.gsub(filter, "#{COLORS[:yellow]}#{filter}#{COLORS[:reset]}")
    elsif regex_pattern
      text.gsub(Regexp.new(regex_pattern), "#{COLORS[:yellow]}\\0#{COLORS[:reset]}")
    else
      # Highlight level
      LEVEL_COLORS.each do |level, color|
        if text.include?(level)
          return text.sub(level, "#{COLORS[color]}#{level}#{COLORS[:reset]}")
        end
      end
      text
    end
  end

  def process_file(filepath, filter: nil, regex_pattern: nil, levels: [], since: nil,
                   context: 0, watch: false, stats: false, no_color: false)
    unless File.exist?(filepath)
      warn "File '#{filepath}' not found."
      return
    end

    if stats
      show_stats(filepath)
      return
    end

    since_sec = parse_since(since || @config['since'] || '1h')
    cutoff = Time.now - since_sec

    lines = File.readlines(filepath, chomp: true)
    entries = lines.map { |line| parse_log_line(line) }

    filtered = []
    entries.each_with_index do |entry, i|
      line = lines[i]
      # Time filter
      if entry[:timestamp] && entry[:timestamp].to_time < cutoff
        next
      end
      # Level filter
      if !levels.empty? && !levels.include?(entry[:level])
        next
      end
      # Pattern filter
      if filter && !line.downcase.include?(filter.downcase)
        next
      end
      if regex_pattern && !line.match?(Regexp.new(regex_pattern))
        next
      end
      filtered << { idx: i, line: line, entry: entry }
    end

    puts "\n📜 Log Viewer – #{filepath}"
    puts "Filter: #{filter || regex_pattern}" if filter || regex_pattern
    puts "Levels: #{levels.join(', ')}" unless levels.empty?
    puts "Watch mode: #{watch ? 'on' : 'off'}\n"

    if context > 0
      shown = Set.new
      filtered.each do |f|
        (-context..context).each do |offset|
          idx = f[:idx] + offset
          shown << idx if idx >= 0 && idx < lines.length
        end
      end
      shown.sort.each do |idx|
        prefix = filtered.any? { |f| f[:idx] == idx } ? '> ' : '  '
        line = lines[idx]
        line = highlight(line, filter, regex_pattern) unless no_color
        puts "#{prefix}#{idx+1.to_s.rjust(4)}: #{line}"
      end
    else
      filtered.each do |f|
        line = f[:line]
        line = highlight(line, filter, regex_pattern) unless no_color
        puts "#{(f[:idx]+1).to_s.rjust(4)}: #{line}"
      end
    end
    puts "\n#{filtered.length} lines matched."

    if watch
      watch_mode(filepath, filter, regex_pattern, levels, since_sec, context, no_color)
    end
  end

  def watch_mode(filepath, filter, regex_pattern, levels, since_sec, context, no_color)
    puts "\n👁️  Watching for new lines (press Ctrl+C to stop)..."
    last_pos = File.size(filepath)
    loop do
      sleep 1
      next unless File.exist?(filepath)
      new_size = File.size(filepath)
      next if new_size <= last_pos
      File.open(filepath, 'r') do |f|
        f.seek(last_pos)
        f.each_line do |line|
          line.chomp!
          entry = parse_log_line(line)
          if entry[:timestamp] && entry[:timestamp].to_time < Time.now - since_sec
            next
          end
          if !levels.empty? && !levels.include?(entry[:level])
            next
          end
          if filter && !line.downcase.include?(filter.downcase)
            next
          end
          if regex_pattern && !line.match?(Regexp.new(regex_pattern))
            next
          end
          line = highlight(line, filter, regex_pattern) unless no_color
          puts "  #{line}"
        end
        last_pos = f.pos
      end
    end
  rescue Interrupt
    puts "\n👋 Watching stopped."
  end

  def show_stats(filepath)
    counts = { 'ERROR' => 0, 'WARN' => 0, 'INFO' => 0, 'DEBUG' => 0, 'OTHER' => 0 }
    File.readlines(filepath).each do |line|
      entry = parse_log_line(line)
      if entry[:level] && counts.key?(entry[:level])
        counts[entry[:level]] += 1
      elsif !line.strip.empty?
        counts['OTHER'] += 1
      end
    end
    puts "\n📊 Log Statistics:"
    ['ERROR', 'WARN', 'INFO', 'DEBUG', 'OTHER'].each do |level|
      color = LEVEL_COLORS[level] || :white
      puts "  #{COLORS[color]}#{level.rjust(5)}#{COLORS[:reset]}: #{counts[level]}"
    end
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: log_filter.rb <file> [options]"
  opts.on("--filter PATTERN", "Filter by keyword") { |v| options[:filter] = v }
  opts.on("--regex PATTERN", "Filter by regex") { |v| options[:regex] = v }
  opts.on("--level LEVELS", "Filter by levels (comma-separated)") { |v| options[:levels] = v }
  opts.on("--since TIME", "Time range (e.g., 5m, 1h, 2d)") { |v| options[:since] = v }
  opts.on("--context N", Integer, "Context lines") { |v| options[:context] = v }
  opts.on("--watch", "Watch for new lines") { options[:watch] = true }
  opts.on("--stats", "Show statistics") { options[:stats] = true }
  opts.on("--no-color", "Disable color") { options[:no_color] = true }
  opts.on("--save", "Save settings") { options[:save] = true }
end.parse!

filepath = ARGV.shift
unless filepath
  puts "Usage: log_filter.rb <file> [options]"
  exit 1
end

app = LogFilter.new

if options[:save]
  levels = options[:levels] ? options[:levels].split(',').map(&:strip) : []
  app.config['filters'] = [options[:filter]] if options[:filter]
  app.config['levels'] = levels
  app.config['since'] = options[:since] || '1h'
  app.save_config
  puts "✅ Filter settings saved."
  exit 0
end

unless options[:filter] || options[:regex] || options[:levels] || options[:since]
  options[:filter] = app.config['filters'][0]
  options[:levels] = app.config['levels'].join(',')
  options[:since] = app.config['since']
end

levels = options[:levels] ? options[:levels].split(',').map(&:strip) : []

app.process_file(filepath,
                 filter: options[:filter],
                 regex_pattern: options[:regex],
                 levels: levels,
                 since: options[:since],
                 context: options[:context] || 0,
                 watch: options[:watch],
                 stats: options[:stats],
                 no_color: options[:no_color])

File: log_filter.js
// log_filter.js
#!/usr/bin/env node
const fs = require('fs');
const readline = require('readline');
const { program } = require('commander');

const CONFIG_FILE = 'log_filter_config.json';

const COLORS = {
    red: '\x1b[91m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    cyan: '\x1b[96m',
    reset: '\x1b[0m'
};

const LEVEL_COLORS = {
    ERROR: 'red',
    WARN: 'yellow',
    INFO: 'green',
    DEBUG: 'cyan'
};

class LogFilter {
    constructor() {
        this.config = this.loadConfig();
    }

    loadConfig() {
        if (fs.existsSync(CONFIG_FILE)) {
            try {
                return JSON.parse(fs.readFileSync(CONFIG_FILE));
            } catch (e) {}
        }
        return { filters: [], levels: [], since: '1h' };
    }

    saveConfig() {
        fs.writeFileSync(CONFIG_FILE, JSON.stringify(this.config, null, 2));
    }

    parseSince(sinceStr) {
        if (!sinceStr) return 3600000;
        const num = parseInt(sinceStr);
        const unit = sinceStr.slice(-1);
        switch (unit) {
            case 'm': return num * 60 * 1000;
            case 'h': return num * 60 * 60 * 1000;
            case 'd': return num * 24 * 60 * 60 * 1000;
            default: return 3600000;
        }
    }

    parseLogLine(line) {
        const patterns = [
            /\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\]\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)/,
            /(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)/,
            /(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\s+(ERROR|WARN|INFO|DEBUG)\s+(.*)/,
            /^(ERROR|WARN|INFO|DEBUG)\s+(.*)/
        ];
        for (const pattern of patterns) {
            const match = line.match(pattern);
            if (match) {
                let dt = null;
                if (match.length === 4) {
                    dt = new Date(match[1]);
                    return { timestamp: isNaN(dt) ? null : dt, level: match[2], message: match[3], raw: line };
                } else if (match.length === 3) {
                    return { timestamp: null, level: match[1], message: match[2], raw: line };
                }
            }
        }
        return { timestamp: null, level: null, message: line, raw: line };
    }

    highlight(text, filter, regexPattern) {
        if (filter) {
            return text.split(filter).join(`${COLORS.yellow}${filter}${COLORS.reset}`);
        }
        if (regexPattern) {
            const re = new RegExp(regexPattern, 'g');
            return text.replace(re, `${COLORS.yellow}$&${COLORS.reset}`);
        }
        // Highlight level
        for (const [level, color] of Object.entries(LEVEL_COLORS)) {
            if (text.includes(level)) {
                return text.replace(level, `${COLORS[color]}${level}${COLORS.reset}`);
            }
        }
        return text;
    }

    processFile(filepath, filter, regexPattern, levels, sinceStr, context, watch, stats, noColor) {
        if (!fs.existsSync(filepath)) {
            console.error(`File '${filepath}' not found.`);
            return;
        }

        if (stats) {
            this.showStats(filepath);
            return;
        }

        const sinceMs = this.parseSince(sinceStr);
        const cutoff = Date.now() - sinceMs;

        const content = fs.readFileSync(filepath, 'utf8');
        const lines = content.split('\n').filter(l => l !== '');
        const entries = lines.map(line => this.parseLogLine(line));

        const filtered = [];
        for (let i = 0; i < entries.length; i++) {
            const entry = entries[i];
            const line = lines[i];
            // Time filter
            if (entry.timestamp && entry.timestamp.getTime() < cutoff) continue;
            // Level filter
            if (levels.length > 0 && !levels.includes(entry.level)) continue;
            // Pattern filter
            if (filter && !line.toLowerCase().includes(filter.toLowerCase())) continue;
            if (regexPattern) {
                const re = new RegExp(regexPattern);
                if (!re.test(line)) continue;
            }
            filtered.push({ idx: i, line, entry });
        }

        console.log(`\n📜 Log Viewer – ${filepath}`);
        if (filter || regexPattern) console.log(`Filter: ${filter || regexPattern}`);
        if (levels.length) console.log(`Levels: ${levels.join(', ')}`);
        console.log(`Watch mode: ${watch ? 'on' : 'off'}\n`);

        if (context > 0) {
            const shown = new Set();
            for (const f of filtered) {
                for (let offset = -context; offset <= context; offset++) {
                    const idx = f.idx + offset;
                    if (idx >= 0 && idx < lines.length) shown.add(idx);
                }
            }
            const sorted = [...shown].sort((a, b) => a - b);
            for (const idx of sorted) {
                const prefix = filtered.some(f => f.idx === idx) ? '> ' : '  ';
                let line = lines[idx];
                if (!noColor) line = this.highlight(line, filter, regexPattern);
                console.log(`${prefix}${String(idx+1).padStart(4)}: ${line}`);
            }
        } else {
            for (const f of filtered) {
                let line = f.line;
                if (!noColor) line = this.highlight(line, filter, regexPattern);
                console.log(`${String(f.idx+1).padStart(4)}: ${line}`);
            }
        }
        console.log(`\n${filtered.length} lines matched.`);

        if (watch) {
            this.watchMode(filepath, filter, regexPattern, levels, sinceMs, context, noColor);
        }
    }

    watchMode(filepath, filter, regexPattern, levels, sinceMs, context, noColor) {
        console.log('\n👁️  Watching for new lines (press Ctrl+C to stop)...');
        let lastPos = fs.statSync(filepath).size;
        const interval = setInterval(() => {
            const stats = fs.statSync(filepath);
            if (stats.size > lastPos) {
                const stream = fs.createReadStream(filepath, { start: lastPos, encoding: 'utf8' });
                const rl = readline.createInterface({ input: stream });
                rl.on('line', (line) => {
                    const entry = this.parseLogLine(line);
                    if (entry.timestamp && entry.timestamp.getTime() < Date.now() - sinceMs) return;
                    if (levels.length > 0 && !levels.includes(entry.level)) return;
                    if (filter && !line.toLowerCase().includes(filter.toLowerCase())) return;
                    if (regexPattern) {
                        const re = new RegExp(regexPattern);
                        if (!re.test(line)) return;
                    }
                    if (!noColor) line = this.highlight(line, filter, regexPattern);
                    console.log(`  ${line}`);
                });
                rl.on('close', () => {
                    lastPos = stats.size;
                });
            }
        }, 1000);
        process.on('SIGINT', () => {
            clearInterval(interval);
            console.log('\n👋 Watching stopped.');
            process.exit(0);
        });
    }

    showStats(filepath) {
        const counts = { ERROR: 0, WARN: 0, INFO: 0, DEBUG: 0, OTHER: 0 };
        const content = fs.readFileSync(filepath, 'utf8');
        for (const line of content.split('\n')) {
            const entry = this.parseLogLine(line);
            if (entry.level && counts[entry.level] !== undefined) {
                counts[entry.level]++;
            } else if (line.trim()) {
                counts['OTHER']++;
            }
        }
        console.log('\n📊 Log Statistics:');
        for (const level of ['ERROR', 'WARN', 'INFO', 'DEBUG', 'OTHER']) {
            const color = LEVEL_COLORS[level] || 'white';
            console.log(`  ${COLORS[color] || ''}${level.padStart(5)}${COLORS.reset}: ${counts[level]}`);
        }
    }
}

program
    .argument('<file>', 'Path to log file')
    .option('--filter <pattern>', 'Filter by keyword')
    .option('--regex <pattern>', 'Filter by regex pattern')
    .option('--level <levels>', 'Filter by log level (comma-separated)')
    .option('--since <time>', 'Time range (e.g., 5m, 1h, 2d)')
    .option('--context <n>', 'Lines of context around matches', parseInt, 0)
    .option('--watch', 'Watch for new lines')
    .option('--stats', 'Show statistics')
    .option('--no-color', 'Disable color output')
    .option('--save', 'Save current filter settings')
    .action((file, options) => {
        const app = new LogFilter();
        let levels = options.level ? options.level.split(',').map(s => s.trim()) : [];

        if (options.save) {
            app.config.filters = [options.filter];
            app.config.levels = levels;
            app.config.since = options.since || '1h';
            app.saveConfig();
            console.log('✅ Filter settings saved.');
            return;
        }

        // Use config defaults if not provided
        if (!options.filter && !options.regex && !options.level && !options.since) {
            options.filter = app.config.filters[0];
            levels = app.config.levels;
            options.since = app.config.since || '1h';
        }

        app.processFile(file, options.filter, options.regex, levels, options.since,
                        options.context, options.watch, options.stats, options.noColor);
    });

program.parse(process.argv);
